lib/versions.ps1

```diff
@@ -1,53 +1,175 @@
-    (manifest $app $bucket $url).version
+function Get-LatestVersion {
+    <#
+    .SYNOPSIS
+        Get latest version of app
+    .DESCRIPTION
+        Get latest version of app from manifest
+    #>
+    param (
+        [String]
+        # App's name
+        $App,
+        [String]
+        # Bucket which the app is belong to
+        $Bucket,
+        [String]
+        # Remote app manifest's URI
+        $URL
+    )
+    return (manifest $App $Bucket $URL).version
 }
-function current_version($app, $global) {
-    @(versions $app $global)[-1]
-}
-function versions($app, $global) {
-    $appdir = appdir $app $global
-    if(!(test-path $appdir)) { return @() }
 
-    sort_versions (Get-ChildItem $appdir -dir -attr !reparsePoint | Where-Object { $null -ne $(Get-ChildItem $_.fullname) } | ForEach-Object { $_.name })
+function Select-CurrentVersion {
+    <#
+    .SYNOPSIS
+        Select current version of app
+    .DESCRIPTION
+        Select current version of installed app, from 'current\manifest.json' or modified time of version directory
+    #>
+    param (
+        [String]
+        # App's name
+        $App,
+        [Switch]
+        # If global installed
+        $Global
+    )
+
+    $appPath = appdir $App $Global
+    if (Test-Path "$appPath\current") {
+        $currentVersion = (installed_manifest $App 'current' $Global).version
+    } else {
+        $installedVersion = Get-InstalledVersion -App $App -Global:$Global
+        if ($installedVersion) {
+            $currentVersion = $installedVersion[-1]
+        } else {
+            $currentVersion = $null
+        }
+    }
+    return $currentVersion
 }
 
-function version($ver) {
-    $ver -split '[\.-]' | ForEach-Object {
-        $num = $_ -as [int]
-        if($num) { $num } else { $_ }
+function Get-InstalledVersion {
+    <#
+    .SYNOPSIS
+        Get installed version of app
+    .DESCRIPTION
+        Get all installed version of app, by checking version directories' 'install.json'
+    #>
+    param (
+        [String]
+        # App's name
+        $App,
+        [Switch]
+        # If global installed
+        $Global
+    )
+
+    $appPath = appdir $App $Global
+    if (!(Test-Path $appPath)) {
+        return @()
+    } else {
+        return @((Get-ChildItem "$appPath\*\install.json" | Sort-Object -Property LastWriteTimeUtc).Directory.Name) -ne 'current'
     }
+
+    # Deprecated
+    # sort_versions (Get-ChildItem $appPath -dir -attr !reparsePoint | Where-Object { $null -ne $(Get-ChildItem $_.FullName) } | ForEach-Object { $_.Name })
 }
-function compare_versions($a, $b) {
-    $ver_a = @(version $a)
-    $ver_b = @(version $b)
 
-    for($i=0;$i -lt $ver_a.length;$i++) {
-        if($i -gt $ver_b.length) { return 1; }
+function Compare-Version {
+    <#
+    .SYNOPSIS
+        Compare versions
+    .DESCRIPTION
+        Compare versions, mainly according to SemVer's rules
+    #>
+    [OutputType('System.Int32')]
```

Comment:
In #3533 you are using bracket notation instead of string. Decide which syntax you want to use and keep it consistent

Reply:
bracket notation should be enough.